numpy/core/arrayprint.py

```diff
@@ -692,7 +694,7 @@ def recurser(index, hanging_indent, curr_width):
         axes_left = a.ndim - axis
 
         if axes_left == 0:
-            return format_function(a[index])
+            return format_function(np.ndarray.__getitem__(a, index))
```

Comment:
Not a huge fan of this - I was hoping to take advantage of this using subclass indexing to enable formatting of masked arrays in future.

Reply:
Right.. masked arrays return an `mvoid` which has the mask information, which would be nice to pass to `format_function`.

One possibility is to make the `format_functions` here convert 0ds to scalars, so the conversion happens inside `format_function` instead of before it. Then we can leave the masked-format_functions as they are. Might be a little ugly though.

Reply:
I think that's the right fix, but I think we can put that off until we work out how to override format functions for subclasses. So I'd be inclined to put the `asarray` at the top level and leave it at that.

Reply:
Hmm, actually putting `asarray` causes the new tests I added in this PR to fail due to infinite recursion on recursive-subtyped-object arrays, eg `x = sub(None); x[()] = x; repr(x)`. The recursive guard fails to catch the recursions.

What happens is `np.asarray` converts the outer array to an ndarray, but the inner element is still the sub type `x`, and so upon recursion `_recursive_guard` does not see that the outer and inner arrays are the same. `np.asanyarray` does not have this problem since it preserves subclass.

I haven't found a better solution than `np.asanyarray` plus `np.__getitem__` for now. Maybe we should merge as is and consider a more involved solution for 1.15.

Reply:
Never mind, I figured out a better place to put the asarray.